mypyc/test-data/run-classes.test

```diff
@@ -1924,3 +1946,263 @@ from native import A, B, C
 a = A()
 b = B()
 c = C()
+
+[case testCopyAlwaysDefinedAttributes]
+import copy
+from typing import Union
+
+class A: pass
+
+class C:
+    def __init__(self, n: int = 0) -> None:
+        self.n = n
+        self.s = ""
+        self.t = ("", 0)
+        self.u: Union[str, bytes] = ''
+        self.a = A()
+
+def test_copy() -> None:
+    c1 = C()
+    c1.n = 1
+    c1.s = "x"
+    c2 = copy.copy(c1)
+    assert c2.n == 1
+    assert c2.s == "x"
+    assert c2.t == ("", 0)
+    assert c2.u == ''
+    assert c2.a is c1.a
+
+[case testNonNativeCallsToDunderNewAndInit]
+from typing import Any
+from testutil import assertRaises
+
+count_c = 0
+
+class C:
+    def __init__(self) -> None:
+        self.x = 'a'  # Always defined attribute
+        global count_c
+        count_c += 1
+
+    def get(self) -> str:
+        return self.x
+
+def test_no_init_args() -> None:
+    global count_c
+    count_c = 0
+
+    # Use Any to get non-native semantics
+    cls: Any = C
+    # __new__ implicitly calls __init__ for native classes
+    obj = cls.__new__(cls)
+    assert obj.get() == 'a'
+    assert count_c == 1
+    # Make sure we don't call __init__ twice
+    obj2 = cls()
+    assert obj2.get() == 'a'
+    assert count_c == 2
+
+count_d = 0
+
+class D:
+    def __init__(self, x: str) -> None:
+        self.x = x  # Always defined attribute
+        global count_d
+        count_d += 1
+
+    def get(self) -> str:
+        return self.x
+
+def test_init_arg() -> None:
+    global count_d
+    count_d = 0
+
+    # Use Any to get non-native semantics
+    cls: Any = D
+    # __new__ implicitly calls __init__ for native classes
+    obj = cls.__new__(cls, 'abc')
+    assert obj.get() == 'abc'
+    assert count_d == 1
+    # Make sure we don't call __init__ twice
+    obj2 = cls('x')
+    assert obj2.get() == 'x'
+    assert count_d == 2
+    # Keyword args should work
+    obj = cls.__new__(cls, x='abc')
+    assert obj.get() == 'abc'
+    assert count_d == 3
+
+def test_invalid_init_args() -> None:
+    # Use Any to get non-native semantics
+    cls: Any = D
+    with assertRaises(TypeError):
+        cls()
+    with assertRaises(TypeError):
+        cls(y='x')
+    with assertRaises(TypeError):
+        cls(1)
+
+[case testTryDeletingAlwaysDefinedAttribute]
+from typing import Any
+from testutil import assertRaises
+
+class C:
+    def __init__(self) -> None:
+        self.x = 0
+
+class D(C):
+    pass
+
+def test_try_deleting_always_defined_attr() -> None:
+    c: Any = C()
+    with assertRaises(AttributeError):
+        del c.x
+    d: Any = D()
+    with assertRaises(AttributeError):
+        del d.x
+
+[case testAlwaysDefinedAttributeAndAllowInterpretedSubclasses]
+from mypy_extensions import mypyc_attr
+
+from m import define_interpreted_subclass
+
+@mypyc_attr(allow_interpreted_subclasses=True)
```

Comment:
Would it make sense to add a test with both `allow_interpreted_subclasses` and `serializable`? For example, what if they are "conflicting" like former is True but latter is False.

Reply:
This would be reasonable. `allow_interpreted_subclasses` would take precedence and the class would be serializable. In the future we could reject incompatible attribute values.